### PR TITLE
fix(no-undef): handle class declarations hoisting

### LINKED ISSUE #463: no-undef false positive
```
...

  function createHttpClient(options) {
    return new HttpClient(opCreateHttpClient(options));
  }

  function opCreateHttpClient(args) {
    return core.jsonOpSync("op_create_http_client", args);
  }

  class HttpClient {
    constructor(rid) {
      this.rid = rid;
    }
    close() {
      core.close(this.rid);
    }
  }


...
```

```
error[no-undef]: HttpClient is not defined
   --> /Users/biwanczuk/dev/deno/op_crates/fetch/26_fetch.js:877:16
    |
877 |     return new HttpClient(opCreateHttpClient(options));
    |                ^^^^^^^^^^
    |
```

Note: the file in question is a script, but it shouldn't really matter here.

### PR DESCRIPTION
**Pull request by magurotuna**
Fixes #463 

This PR fixes #463 by adding a new scope analyzer named `DeclFinder`. Although there's still a problem with variables that are declared with `var`, it works fine for Deno's runtime JS scripts.

### CODE REVIEW

**Comment by magurotuna** [comment]
Current status: I've completed fixing the false positive described in #463. Now, running `deno_lint` for Deno's runtime JS code, we can see two discrete errors about `no-undef`:

```
  17   │ error[no-undef]: bootstrap is not defined
  18   │  --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/90_deno_ns.js:3:1
  19   │   |
  20   │ 3 | __bootstrap.denoNs = {
  21   │   | ^^^^^^^^^^^
  22   │   |
  23   │ error[no-undef]: bootstrap is not defined
  24   │  --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/90_deno_ns.js:4:9
  25   │   |
  26   │ 4 |   test: __bootstrap.testing.test,
  27   │   |         ^^^^^^^^^^^
  28   │   |
  29   │ error[no-undef]: bootstrap is not defined
  30   │  --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/90_deno_ns.js:5:12
  31   │   |
  32   │ 5 |   metrics: __bootstrap.metrics.metrics,
  33   │   |            ^^^^^^^^^^^
  34   │   |
  35   │ error[no-undef]: bootstrap is not defined
  36   │  --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/90_deno_ns.js:6:12
  37   │   |
  38   │ 6 |   Process: __bootstrap.process.Process,
  39   │   |            ^^^^^^^^^^^
  40   │   |

.....(the rest is omitted)
```

`90_deno_ns.js` has many errors about `__bootstrap`.

---------------------

```
 843   │ error[no-undef]: currentPart is not defined
 844   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1161:5
 845   │      |
 846   │ 1161 |     currentPart = "";
 847   │      |     ^^^^^^^^^^^
 848   │      |
 849   │ error[no-undef]: currentPart is not defined
 850   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1172:25
 851   │      |
 852   │ 1172 |           const value = currentPart.trim();
 853   │      |                         ^^^^^^^^^^^
 854   │      |
 855   │ error[no-undef]: currentPart is not defined
 856   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1177:11
 857   │      |
 858   │ 1177 |           currentPart = "";
 859   │      |           ^^^^^^^^^^^
 860   │      |
 861   │ error[no-undef]: currentPart is not defined
 862   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1182:22
 863   │      |
 864   │ 1182 |         currentKey = currentPart.trim();
 865   │      |                      ^^^^^^^^^^^
 866   │      |
 867   │ error[no-undef]: currentPart is not defined
 868   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1183:9
 869   │      |
 870   │ 1183 |         currentPart = "";
 871   │      |         ^^^^^^^^^^^
 872   │      |
 873   │ error[no-undef]: currentPart is not defined
 874   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1187:7
 875   │      |
 876   │ 1187 |       currentPart += c;
 877   │      |       ^^^^^^^^^^^
 878   │      |
 879   │ error[no-undef]: currentPart is not defined
 880   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1190:21
 881   │      |
 882   │ 1190 |       const value = currentPart.trim();
 883   │      |                     ^^^^^^^^^^^
 884   │      |
 885   │ error[no-undef]: currentPart is not defined
 886   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/cli/rt/02_console.js:1195:7
 887   │      |
 888   │ 1195 |       currentPart = "";
 889   │      |       ^^^^^^^^^^^
 890   │      |
```

Looking at `02_console.js`, we can find `currentPart` actually not defined. So this is "true positive".

**Comment by magurotuna** [comment]
However current implementation doesn't handle variables declared with `var` properly. I have to fix it.

File: src/rules/no_undef.rs [hunk 1]
```diff
@@ -26,139 +23,809 @@ impl LintRule for NoUndef {
...
+    //       {
+    //         var target = 0;
+    //       }
+    //     }
+    //     "#;
+    //   let idents = get_idents(src, "target");
+    //   let finder = decl_finder(src);
+    //   let target_used = &idents[0];
+    //   assert!(finder.decl_exists(&target_used));
+    // }
```
Comment (magurotuna) [inline src/rules/no_undef.rs:807]: This example shows that `DeclFinder` (which is `no-undef` specific analyzer) cannot deal with `var`. If we uncomment this test, test will fail.
`var target = 0;` is not block-scoped but function-scoped, so `target++` should be treated as valid.

CC @bartlomieju @kdy1

Comment (kdy1) [reply src/rules/no_undef.rs:807]: This seems like a bug of resolver in swc. I'll check if it's the case.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: Oh really? I think there's a problem with my implementation of `DeclFinder`. In `DeclFinder`, it collects variable declarations in a naive way. It doen't use swc's `SyntaxContext` or something like that.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: To be honest I have no idea what `SyntaxContext` exactly is. In the original implementation of `no-undef`, the following code resulted in false positive, which was perhaps caused by `SyntaxContext` problem, I guess.

```ts
(() => {
  function foo() {
    return new Bar();
  }
  class Bar {}
})();
```

Comment (kdy1) [reply src/rules/no_undef.rs:807]: It was not, while the code in #463 is bug of the swc resolver.
I'll investigate further.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: I see. Thanks for your help :)

Comment (kdy1) [reply src/rules/no_undef.rs:807]: Does deno_lint does some transforms?
Root cause of the failing test is `visit_var_declarator` not being called.

Comment (kdy1) [reply src/rules/no_undef.rs:807]: Oh, I did wrong test.

Comment (kdy1) [reply src/rules/no_undef.rs:807]: Changing to
```rust

impl BindingCollector {
  fn declare(&mut self, i: Id) {
    dbg!(&i);
    self.declared.insert(i);
  }
}

impl Visit for BindingCollector {
  fn visit_fn_decl(&mut self, f: &FnDecl, _: &dyn Node) {
    dbg!(&f.ident);
    self.declare(f.ident.to_id());
    f.visit_children_with(self);
  }
}
```

fixes the issue. (I changed it to visit children nodes)

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: Great! I'll give it a try

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: hmm... tried your fix but the test still fails.

Code fix:
https://github.com/magurotuna/deno_lint/blob/test-magurotuna/src/rules/no_undef.rs#L56-L66

Test case:
https://github.com/magurotuna/deno_lint/blob/test-magurotuna/src/rules/no_undef.rs#L352-L368

Test result:

```
thread 'rules::no_undef::tests::no_undef_valid' panicked at 'Unexpected diagnostics found:
[
    LintDiagnostic {
        range: Range {
            start: Position {
                line: 4,
                col: 19,
                byte_pos: 56,
            },
            end: Position {
                line: 4,
                col: 22,
                byte_pos: 59,
            },
        },
        filename: "deno_lint_test.tsx",
        message: "Bar is not defined",
        code: "no-undef",
        hint: None,
    },
]

source:

    (() => {
      function foo() {
        return new Bar();
      }
      class Bar {}
    })();

', src/test_util.rs:228:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Do I miss something?

Comment (kdy1) [reply src/rules/no_undef.rs:807]:
```js
(() => {
      function foo() {
        return new Bar();
      }
      class Bar {}
    })();
```

This is bug of swc resolver. I'm currently working on it. I think the test below is a same bug.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: > This is bug of swc resolver. I'm currently working on it. I think the test below is a same bug.This is bug of swc resolver. I'm currently working on it. I think the test below is a same bug.

Oh I see. Then I'll wait for swc resolve to be fixed 👍

Comment (kdy1) [reply src/rules/no_undef.rs:807]: I fixed it and published a new version of `swc_ecmascript` and `swc_ecma_transforms`.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: @kdy1 
Thanks for the quick fix!
`deno_lint` depends on `swc_ecmascript`, doesn't depend on `swc_scma_transforms` directly. It seems like `swc_ecma_transforms` as a dependency of `swc_ecmascript` was not updated. (in https://github.com/swc-project/swc/blob/master/ecmascript/Cargo.toml#L27, the version of transforms is still `0.29.1`)
Is this correct?

Comment (kdy1) [reply src/rules/no_undef.rs:807]: Oh. I forgot updating it but you can do `cargo update -p swc_ecma_transforms`.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: nice command! That allows to pass the test. 
I have one more question. Now we've added `f.visit_children_with(self);` to `visit_fn_decl`. Should we add `visit_children_with` to the other visit methods e.g. `visit_class_decl`? I feel like this is required to handle a case like the following:

```ts
class SomeClass {
  someMethod() {
    function foo() {
      return new Bar();
    }
    class Bar {}
  }
}
```

Comment (kdy1) [reply src/rules/no_undef.rs:807]: Yes. The binding collector should collect all bindings.
Alternatively, you can change it to `VisitAll`.

Comment (magurotuna) [reply src/rules/no_undef.rs:807]: oh, right, `VisitAll` is the finest solution here. I'll use that :)

**Comment by magurotuna** [comment]
Closing in favor of #497

**Pull request closed without merging**